Sources/ProjectSpec/ProjectSpec.swift

```diff
@@ -0,0 +1,90 @@
+import Foundation
+import typealias JSONUtilities.JSONDictionary
+import PathKit
+
+extension Project {
+
+    public struct Spec {
+        public let relativePath: Path
+        public let jsonDictionary: JSONDictionary
+        public let subSpecs: [Spec]
+
+        public init(relativePath: Path, jsonDictionary: JSONDictionary, subSpecs: [Spec] = []) {
+            self.relativePath = relativePath
+            self.jsonDictionary = jsonDictionary
+            self.subSpecs = subSpecs
+        }
+
+        public init(filename: String, basePath: Path, relativePath: Path = "") throws {
+            let path = basePath + relativePath + filename
+
+            // Depending on the extension we will either load the file as YAML or JSON
+            var json: [String: Any]
+            if path.extension?.lowercased() == "json" {
+                let data: Data = try path.read()
+                let jsonData = try JSONSerialization.jsonObject(with: data, options: .allowFragments)
+                guard let jsonDictionary = jsonData as? [String: Any] else {
+                    fatalError("Invalid JSON at path \(path)")
+                }
+                json = jsonDictionary
+            } else {
+                json = try loadYamlDictionary(path: path)
+            }
+
+            var includeStrings: [String]
+            if let includeString = json["include"] as? String {
+                includeStrings = [includeString]
+            } else if let includeArray = json["include"] as? [String] {
+                includeStrings = includeArray
+            } else {
+                includeStrings = []
+            }
+
+            let includes = try includeStrings.map { include -> Spec in
+                let path = Path(include)
+                return try Spec(filename: path.lastComponent, basePath: basePath + relativePath, relativePath: path.parent())
+            }
+
+            self.relativePath = relativePath
+            self.jsonDictionary = json
+            self.subSpecs = includes
+        }
+
+        public func resolvedDictionary() -> JSONDictionary {
+            return jsonDictionary.merged(onto:
+                subSpecs
+                    .map { $0.resolvedDictionary() }
+                    .reduce([:]) { $1.merged(onto: $0) }
+            )
+        }
+    }
+}
+
+internal protocol SubSequenceConvertible: Sequence {
+
+    init(_ subsequence: SubSequence)
+}
+
+extension String: SubSequenceConvertible { }
+extension Substring: SubSequenceConvertible { }
+
+extension Dictionary where Key: StringProtocol, Key: SubSequenceConvertible, Value: Any {
```

Comment:
Can you explain the reason for the `SubSequenceConvertible` protocol?

Reply:
It's to do with `String` and `Substring`, and stripping the `:REPLACE` suffix from the keys. If the `Dictionary` extension is bound to `StringProtocol: Any`, and you perform this:

```swift
let newKey = key[key.startIndex ..< key.index(key.endIndex, offsetBy: -8)]
merged[newKey] = value
```

You get a compiler error:

> Cannot subscript a value of type '[Dictionary<Key, Value>.Key : Dictionary<Key, Value>.Value]' (aka 'Dictionary<Key, Value>') with an index of type 'Key.SubSequence'

The way to convert a `Substring` to a `String` is simply `String(subtring)`. The `SubSequenceConvertible` protocol just describes that conversion and makes the above code work via:

```swift
let newKey = key[key.startIndex ..< key.index(key.endIndex, offsetBy: -8)]
merged[Key(newKey)] = value
```

Note that it could have been avoided by making the extension `where Key == String, Value: Any`